### ISSUE TITLE
[feature request] Update module parameters after __init__()

### ISSUE THREAD

**Issue by emilmelnikov** [pytorch/pytorch#7620]
Currently, modules assign their parameters within constructors (see [Linear](https://github.com/pytorch/pytorch/blob/c425d0350b9d4a4891eeb47c06ace1efb53da508/torch/nn/modules/linear.py#L41) and [_ConvNd](https://github.com/pytorch/pytorch/blob/c425d0350b9d4a4891eeb47c06ace1efb53da508/torch/nn/modules/conv.py#L29) `__init__()` source). This makes _ad-hoc_ patching of already created models not as flexible as it could be.

As a motivating use case, one may want to adapt pre-packaged ResNet-18 for MNIST-like data. An intuitive solution would be the following:
```python
model = resnet18()
model.conv1.in_channels = 1
model.fc.out_features = 10
# New API: recreate contents of _parameters and _buffers
# based on the modules' arguments supplied in __init__().
model.refresh_parameters()
```

It seems that the current solution is [to assign a fresh instance of the same module to it's parent](https://discuss.pytorch.org/t/how-to-modify-the-final-fc-layer-based-on-the-torch-model/766/3). However, one needs to pass all other module arguments. This gets especially tedious with convs (stride, kernel size, paddings etc.).

Another solution is to add a `copy(arg1=foo, arg2=bar)` method to all modules that clones the module with the updated arguments `arg1` and `arg2`, set to `foo` and `bar`. This way, the example above would look like this:
```python
model = resnet18()
model.conv1 = model.conv1.copy(in_channels=1)
model.fc.out_features = model.fc.copy(out_features=10)
```

**Comment by SsnL** [comment]
`out_features` defines the shape of the parameter tensor. I'm not sure how you can change it.

**Comment by emilmelnikov** [comment]
Yes, it's assigned in the `__init__` now:
```python
def __init__(self, in_features, out_features, bias=True):
    super(Linear, self).__init__()
    self.in_features = in_features
    self.out_features = out_features
    self.weight = Parameter(torch.Tensor(out_features, in_features))
    if bias:
        self.bias = Parameter(torch.Tensor(out_features))
    else:
        self.register_parameter('bias', None)
    self.reset_parameters()
```

The original proposal was to extract initialization of `self.weight` and `self.bias` into separate method.
Something like this:
```python
def __init__(self, in_features, out_features, bias=True):
    super(Linear, self).__init__()
    self.in_features = in_features
    self.out_features = out_features
    self.use_bias = bias
    self.refresh_parameters()

def refresh_parameters(self):
    del self._parameters['weight']
    del self._parameters['bias']
    self.weight = Parameter(torch.Tensor(self.out_features, self.in_features))
    if self.use_bias:
        self.bias = Parameter(torch.Tensor(self.out_features))
    else:
        self.register_parameter('bias', None)
    self.reset_parameters()
```

Alternatively, create a special cloning method with optional parameters, like this:

```python
def __init__(self, in_features, out_features, bias=True):
    super(Linear, self).__init__()
    self.in_features = in_features
    self.out_features = out_features
    self.weight = Parameter(torch.Tensor(out_features, in_features))
    if bias:
        self.bias = Parameter(torch.Tensor(out_features))
    else:
        self.register_parameter('bias', None)
    self.reset_parameters()

def copy(self, **kwargs):
    args = dict(
        in_features=self.in_features,
        out_features=self.out_features,
        bias=self.use_bias,
    )
    args.update(kwargs)
    return Linear(**args)
```

**Comment by karandwivedi42** [comment]
Maybe the `nn.functional` interface is more suitable for your use-case? It's an amazing interface when you want to play around with the parameters.

**Comment by emilmelnikov** [comment]
@karandwivedi42 But how the functional interface can be used if you already have a prebuilt `Module`? As far as I understand, in order to work correctly, Modules require Parameters assigned to them.

**Comment by emilmelnikov** [comment]
Another possible way is to create an `arguments` property/method that would return all arguments that define the Module:

```python
class Module:
    @property
    def arguments(self):
        return {}

class Linear:
    @property
    def arguments(self):
        return dict(
            in_features=self.in_features,
            out_features=self.out_features,
            bias=self.bias is not None,
        )
```

Then, solution to the problem becomes trivial:

```python
model = resnet18()
model.fc = Linear({**model.fc.arguments, **dict(out_features=10)})
```

Moreover, I guess this property can be used in other places (e.g. generic `repr` implementations).

**Comment by apaszke** [comment]
I actually like the `arguments` idea, although I would make it `raise NotImplementedError` instead of silently returning an empty dict by default.

**Comment by emilmelnikov** [comment]
Yes, raising an exception is probably a better behavior, but wouldn't this break user-defined modules that don't know or want `arguments`?

**Comment by apaszke** [comment]
I can't see why would that break user defined code? It won't cause an error unless you actually use that attribute. More, if you do use it without defining it first, it should be an error.

**Comment by emilmelnikov** [comment]
Yes, you are correct, sorry. Would you mind if I start working on this feature?

**Comment by emilmelnikov** [comment]
There is an implementation strategy that seems to be consistent with the existing PyTorch design. What if we adopt the parameters/buffers/modules approach and assign arguments in the `Module.__setattr__`? We can either pick up everything that is not a proper parameter/buffer/module, or create a wrapper `Argument` type, piggybacking on the `Parameter` design.

**Comment by emilmelnikov** [comment]
There is another API design issue: how to ensure that an indirect child of `Module` overrides the `arguments` property?

1. We have `class Module(object)` without any arguments.
2. Then, there is `class Threshold(Module)` with arguments `threshold`, `value` and `inplace`.
3. Next, there is `class ReLU(Threshold)` with the only argument `inplace`.

If we accidentally forget to override `parameters` in the ReLU, it will return invalid arguments.

**Comment by apaszke** [comment]
About your second concern: there is a way to do it with metaclasses, but it's too complicated. I don't think we should handle that.

**Comment by emilmelnikov** [comment]
I've implemented this feature, albeit with a different API. It is a bit complex, but hopefully more convenient for module implementors.
I didn't create a PR because tests are not done yet, and I'm not sure whether the proposed API is OK.
Could someone [review the diff](https://github.com/emilmelnikov/pytorch/commit/bf610c8e29d8eb4b3a67fe96b30e7384a99da187#diff-527b04fec26e044c060ad28b22e842beR872) and give some feedback?
